[Add] EcomRLVE Blog (#3307)
huggingface/blog: ecom-rlve.md

---
title: "Ecom-RLVE: Adaptive Verifiable Environments for E-Commerce Conversational Agents"
thumbnail: /blog/assets/ecom-rlve/thumbnail.png
authors:
- user: thebajajra
  org: owlgebra-ai
- user: ai-queen
  org: owlgebra-ai
- user: pmonad
  org: owlgebra-ai
- user: burtenshaw
---

# Ecom-RLVE: Adaptive Verifiable Environments for E-Commerce Conversational Agents

> **TL;DR** — We extend the RLVE framework from single-turn reasoning puzzles to **multi-turn, tool-augmented e-commerce conversations**. EcomRLVE-GYM provides 8 verifiable environments — product discovery, substitution, cart building, returns, order tracking, policy QA, bundle planning, and multi-intent journeys — each with procedural problem generation, a 12-axis difficulty curriculum, and algorithmically verifiable rewards. We train a Qwen 3 8B model with DAPO over 300 steps and present early results demonstrating that environment scaling and adaptive difficulty transfer to agentic, real-world task completion.

> This project originated in the [Pytorch OpenEnv Hackathon](https://cerebralvalley.ai/e/openenv-hackathon-sf) and is still evolving, follow us for updates 🔥 

Reviewer: ```suggestion

This project originated in the [Pytorch OpenEnv Hackathon](https://cerebralvalley.ai/e/openenv-hackathon-sf) and is still evolving, follow us for updates 🔥 
```

---

> 

Reviewer: ```suggestion
```

### From RLVE-Gym to EcomRLVE-GYM

RLVE-Gym provides 400 environments for sorting, multiplication, Sudoku, and other algorithmic-reasoning tasks; however, those are all **single-turn, text-in / text-out** puzzles — extending to agentic domains was left as future work.

EcomRLVE-GYM fills that gap: we stay in the **verifiable** regime (e-commerce outcomes *can* be checked algorithmically) while extending to **multi-turn, tool-augmented, agentic** conversations — environments where the agent must *act* (call tools, modify world state) rather than merely *reason* (produce a text answer) and compensates for the deficiency of the search system.

EcomRLVE-GYM transforms customer-service outcomes structurally verifiable:

![verifiable_signals_dark](https://cdn-uploads.huggingface.co/production/uploads/6893dd21467f7d2f5f358a95/dA0i6ZB3JDG-rqQtLRCy0.png)

Every signal above can be evaluated by a program with access to the hidden ground-truth goal. No human annotation or LLM-as-a-judge is needed.

---

## What a training episode looks like

> Before we explain the framework, here is what a single EcomRLVE episode looks like at `d = 4`. The environment generates a hidden goal, a simulated user opens the chat, and the agent must use tools to satisfy the request. Every action is verified algorithmically — no LLM judge required.

Reviewer: ```suggestion
Before we explain the framework, here is what a single EcomRLVE episode looks like at difficulty `d = 4`. The environment generates a hidden goal, a simulated user opens the chat, and the agent must use tools to satisfy the request. Every action is verified algorithmically — no LLM judge required.
```

(difficulty not introduced yet. We can also remove and just say "here's what a single episode looks like", I don't think it's necessary at this point).

<center><img src="https://cdn-uploads.huggingface.co/production/uploads/6893dd21467f7d2f5f358a95/qDXS-CPl8DT4JN6Uq6nrt.png", width="300", height="400", alt="Sample Episode" />
</center>

The reward is fully computed by code: F1 over `(product, variant, qty)` tuples, an efficiency bonus for finishing in fewer turns, and a hallucination check that every recommended product ID was actually retrieved. If the agent had picked the Lightning variant instead of USB-C, the simulated user would have corrected it mid-dialogue — and the F1 would have dropped.

---

---

> 

Reviewer: ```suggestion
```

| **Policy QA** | Answer a deterministic question about store policy (return window, shipping rules, etc.) |
| **Bundle Planning** | Recommend a complete shopping list for a project within a budget |
| **Multi-Intent Journey** | Handle a conversation that chains 2–5 of the above tasks in sequence |

Every environment uses the same three-part reward signal:

- **Task reward** — did the agent actually complete the goal? (e.g., were the right products recommended, was the cart correct, was the right order tracked?)
- **Efficiency reward** — did the agent complete it without wasting turns? Turns the *user* caused (asking a follow-up, confirming an action) don't count against the agent — only turns caused by agent mistakes do.
- **Hallucination penalty** — did the agent only recommend products it actually retrieved during the session? Recommending product IDs that were never looked up is penalised, so the agent cannot invent results from memory.

Invalid outputs (malformed JSON, illegal tool calls) trigger an immediate failure score, creating a strong incentive for well-formed responses from step one.

---

> 

Reviewer: ```suggestion
```

![Screenshot 2026-03-08 at 11.27.11](https://cdn-uploads.huggingface.co/production/uploads/6893dd21467f7d2f5f358a95/SALZRvBC6TP1HG1ZxqWsh.png)

Here are four representative difficulty axes:

| What changes | Easy (`d = 0`) | Medium (`d = 6`) | Hard (`d = 12`) |
|---|---|---|---|
| **How many constraints** the user has | 2 | 5 | 8 |
| **How often the user omits** a constraint | 5% | 70% | ~80% |
| **Fraction of search results** that are distractors | 0% | 12% | 24% |
| **Items that go out of stock** mid-conversation | 0% | 30% | 50% |

The other eight axes cover turn budget, input noise (typos, slang), context switches, retrieval depth, order-history size, policy complexity, and tool budget. The full breakdown is in the [technical report](https://github.com/owlgebra-ai/EcomRLVE-Gym).

**Adaptive scheduling.** Each environment tracks the agent's success rate independently and only advances to harder problems once the agent is passing the current level reliably. This keeps every environment training at the agent's capability frontier — avoiding both "too easy to learn from" and "too hard to make progress on".

---

> 

Reviewer: ```suggestion
```

| **Turn 6** | | Re-fetches tank top variants *again* — **still ignores the air fryer**. User: *"Please change the Air Fryer variant."* |
| **Turn 7** | | *"I cannot find a Charcoal filter variant."* — It exists; the agent saw it in turn 3 but never selected it. User: *"Skip the Air Fryer then."* |
| **Turn 8** | | `answer: { "done": true }` — Cart still wrong: Charger ✓, Air Fryer (Bamboo) ✗, Tank Top (3×XL + 2×XS) ✗ |
| **Reward** | `r_task` = +1.00, `r_eff` = +0.33, `r_hall` = 0.00, **r_total = +0.80** ✓ | `r_task` ≈ 0.00, `r_eff` = −0.43, `r_hall` = 0.00, **r_total = −0.06** ✗ |
| **Outcome** | Cart matches goal. 3 turns, 2 effective. | Wrong variants, wrong quantities, user gave up. 8 turns, 6 effective. |

At d=1 the agent solves the task in 3 clean turns. At d=8 it spirals — picking Bamboo instead of Charcoal, XL instead of XS, never fixing the air fryer despite two user corrections, then hallucinating that the variant doesn't exist. This is exactly the kind of multi-step error cascade that the difficulty curriculum surfaces, and that adaptive training should teach the agent to recover from.

---

> 

Reviewer: ```suggestion
```

## User simulation

A verifiable environment needs a user simulator that behaves realistically. We use **Qwen3.5 (9.7B)** to generate natural, varied user messages rather than canned templates — covering everything from typo-filled requests to mid-conversation topic switches.

Two design choices matter for training quality:

**Preferences match stated constraints.** Each simulated user has a hidden set of preferences (price sensitivity, brand loyalty, shipping speed, etc.). These are deliberately biased toward whatever constraints the user communicated — so if the user said "under \$25", the reward function actually cares about price. Without this, an agent could be penalised for correctly following the user's instructions.

**Strategic omission.** The LLM deliberately withholds some constraints from the opening message to force the agent to ask clarifying questions. The system tracks exactly what was and wasn't mentioned, so the agent is never penalised for information it was never given.

---

> 

Reviewer: ```suggestion
```

## Environment scaling

Following RLVE's methodology, we define nested environment collections:

**C1 ⊂ C2 ⊂ C4 ⊂ C8**

| Collection | Environments | Skills trained |
|------------|-------------|----------------|
| **C1** | Cart | Serarch Query Formulation, Cart Manipulation |
| **C2** | + Substitution | Similarity reasoning under constraints |
| **C4** | + Product Discovery, Returns | Transactional workflows (Retrieval + recommendation, return initiation) |
| **C8** | + Status, Policy, Bundle, Journey | Knowledge retrieval, planning, compositionality |

We hypothesise — consistent with RLVE's findings — that C8 agents outperform single-environment specialists, even on the specialist's own task.

---

> 

Reviewer: ```suggestion
```

## Early results

We trained Qwen 3 8B with DAPO on C1 (Cart Building) for 300 steps as an initial viability study.

| | Config |
|---|--------|
| **Base model** | Qwen 3 8B |
| **Algorithm** | DAPO (G = 8 rollouts/prompt) |
| **LR** | 1e-5 |
| **Catalog** | 2M products, FAISS index with `Alibaba-NLP/gte-modernbert-base` (768-dim) |
| **User sim** | Qwen3.5 9.7B |

![accuracy_levels](https://cdn-uploads.huggingface.co/production/uploads/6893dd21467f7d2f5f358a95/eWQqFP-PbCJeNsn8klCQZ.png)

We saw progressive growth in difficulty reached, confirming that adaptive scheduling produces a steady learning signal rather than the saturation (static-low) or starvation (static-high) patterns predicted by the RLVE paper.

---

> 

Reviewer: ```suggestion
```

## Try it yourself

Run a live episode directly in your browser using the embedded demo below. Here is how to get started:

1. **Pick an environment** from the dropdown (e.g., `E_CART` for cart building or `E_PD` for product discovery).
2. **Set a difficulty** — `0` is a simple single-constraint task; `6+` introduces missing information, noisy retrieval, and variant selection.
3. **Click "Reset Episode"** — the simulated user will open with a shopping request.
4. You are the agentnow: Make tool calls, analyse outputs and submit the final list of product ids.
5. Click **"Reset Episode"** between runs to start a fresh scenario.

<script
	type="module"
	src="https://gradio.s3-us-west-2.amazonaws.com/4.36.1/gradio.js"
></script>

<gradio-app theme_mode="dark" space="owlgebra-ai/EcomRLVE-Gym"></gradio-app>

---

> 

Reviewer: ```suggestion
```

## Resources

[![Models](https://img.shields.io/badge/🤗%20Models-WUFUS-blue)](https://huggingface.co/collections/owlgebra-ai/wufus)
[![Data](https://img.shields.io/badge/🤗%20Catalog%20Data-Amazebay2M-yellow)](https://huggingface.co/datasets/owlgebra-ai/Amazebay-catalog-2M)
[![Code](https://img.shields.io/badge/Github-Code-black)](https://github.com/owlgebra-ai/EcomRLVE-Gym)
[![Demo](https://img.shields.io/badge/🤗-Space-red)](https://huggingface.co/spaces/owlgebra-ai/EcomRLVE-Gym)

The environments, verifiers, and training configs are all open-source:

```bash
git clone https://github.com/owlgebra-ai/EcomRLVE-Gym
cd EcomRLVE-Gym
pip install -e .
```

The 2M-product catalog is on the Hub:

```python
from datasets import load_dataset

catalog = load_dataset("owlgebra-ai/Amazebay-catalog-2M", split="train")
print(f"{len(catalog)} products loaded")
```

---

> 

Reviewer: ```suggestion
```

---
title: "Ecom-RLVE: Adaptive Verifiable Environments for E-Commerce Conversational Agents"
thumbnail: /blog/assets/ecom-rlve/thumbnail.png
authors:
- user: thebajajra
  guest: true
  org: owlgebra-ai
- user: ai-queen
  guest: true
  org: owlgebra-ai
- user: pmonad
  guest: true
  org: owlgebra-ai
- user: burtenshaw
---

# Ecom-RLVE: Adaptive Verifiable Environments for E-Commerce Conversational Agents

[![Models](https://img.shields.io/badge/🤗%20Models-WUFUS-blue)](https://huggingface.co/collections/owlgebra-ai/wufus)
[![Data](https://img.shields.io/badge/🤗%20Catalog%20Data-Amazebay2M-yellow)](https://huggingface.co/datasets/owlgebra-ai/Amazebay-catalog-2M)
[![Code](https://img.shields.io/badge/Github-Code-black)](https://github.com/owlgebra-ai/EcomRLVE-Gym)

> [![Demo](https://img.shields.io/badge/🤗-Space-red)](https://huggingface.co/spaces/owlgebra-ai/EcomRLVE-Gym)

Reviewer: I'd maybe move the resources to the end of the post so readers get the tldr first.

Author: moved to resources

> > **TL;DR** — We extend the RLVE framework ([Zeng et al., 2025](https://arxiv.org/abs/2511.07317)) from single-turn reasoning puzzles to **multi-turn, tool-augmented e-commerce conversations**. EcomRLVE-GYM provides 8 verifiable environments — product discovery, substitution, cart building, returns, order tracking, policy QA, bundle planning, and multi-intent journeys — each with procedural problem generation, a 12-axis difficulty curriculum, and algorithmically verifiable rewards. We train a Qwen 3 1.7B model with DAPO over 300 steps and present early results demonstrating that environment scaling and adaptive difficulty transfer to agentic, real-world task completion.

Reviewer: ```suggestion
> **TL;DR** — We extend the RLVE framework from single-turn reasoning puzzles to **multi-turn, tool-augmented e-commerce conversations**. EcomRLVE-GYM provides 8 verifiable environments — product discovery, substitution, cart building, returns, order tracking, policy QA, bundle planning, and multi-intent journeys — each with procedural problem generation, a 12-axis difficulty curriculum, and algorithmically verifiable rewards. We train a Qwen 3 1.7B model with DAPO over 300 steps and present early results demonstrating that environment scaling and adaptive difficulty transfer to agentic, real-world task completion.
```

I'd skip the citation here to make it a bit less intimidating, and perhaps explain what it is at a later point.
It could also be interesting to provide some intuition on why we are doing this and what's the end goal.

> The project is evolving, started with [Pytorch OpenEnv Hackathon](https://cerebralvalley.ai/e/openenv-hackathon-sf).

Reviewer: ```suggestion
This project originated in the [Pytorch OpenEnv Hackathon](https://cerebralvalley.ai/e/openenv-hackathon-sf) and is still evolving, follow us for updates 🔥 
```

---

## What a training episode looks like

Before we explain the framework, here is what a single EcomRLVE episode looks like at `d = 4`. The environment generates a hidden goal, a simulated user opens the chat, and the agent must use tools to satisfy the request. Every action is verified algorithmically — no LLM judge required.

<center><img src="https://cdn-uploads.huggingface.co/production/uploads/6893dd21467f7d2f5f358a95/qDXS-CPl8DT4JN6Uq6nrt.png", width="300", height="400", alt="Sample Episode" />
</center>

The reward is fully computed by code: F1 over `(product, variant, qty)` tuples, an efficiency bonus for finishing in fewer turns, and a hallucination check that every recommended product ID was actually retrieved. If the agent had picked the Lightning variant instead of USB-C, the simulated user would have corrected it mid-dialogue — and the F1 would have dropped.

> ---

Reviewer: I'd move this to a later state, it's too abrupt to get started with imo.

Author: modified

> ## Why RL for shopping agents?

Reviewer: This is a great intro, I'd open with this section.

Author: moved it as intro

Large language models can hold fluent conversations, yet deploying them as shopping assistants reveals a persistent gap: **fluency ≠ task completion**. A customer who asks *"find me a USB-C charger under \$25 that ships in two days"* needs an agent that invokes the right catalog search, filters on three hard constraints, avoids hallucinating product IDs it never retrieved, and handles follow-ups when the top result goes out of stock.

Supervised fine-tuning can teach surface-level tool use from demonstrations, but it cannot scale to the combinatorial space of constraint configurations, partial-information dialogues, and multi-step transactional workflows that real e-commerce demands.

Reinforcement learning with verifiable rewards (RLVR) offers an alternative: the agent optimises for *outcomes* — did the products satisfy the constraints? Was the cart correct? Was the return initiated for the right order line? The challenge is constructing reward functions that are both **verifiable** (no LLM-as-a-judge subjectivity) and **adaptive** (difficulty that grows with the policy's capability).

### From RLVE-Gym to EcomRLVE-GYM

> [RLVE](https://arxiv.org/abs/2511.07317) (Zeng et al., 2025) introduced adaptive verifiable environments and built RLVE-Gym — 400 environments for sorting, multiplication, Sudoku, and other algorithmic-reasoning tasks. But those are all **single-turn, text-in / text-out** puzzles. The paper's own future-work section calls for extending to agentic domains.

Reviewer: Consider skipping paper citations, I'd suggest to move them to a "resources" section at the end. The main goal of the blog post would be readability (in my opinion), rather than scientific accuracy or comprehensive citations. In this case, for example, we could say something like "RLVE-Gym provides ...; however, ..."

Author: done

Every environment shares a **composite reward**:

```
r = clip(0.75 * r_task + 0.15 * r_eff + 0.10 * r_hall, -1, 1)
```

with hard-fail override (`r = -1`) for invalid JSON, illegal tool calls, or safety violations.

**Fair efficiency scoring with UserActs.** A naive turn-count penalty punishes the agent for every turn — including ones the *user* caused. To fix this, the user simulator tags each response with a structured dialogue act:

| UserAct | Meaning | Penalised? |
|---------|---------|-----------|
| `confirm` | User confirms the agent's action | No |
| `clarify` | User provides previously omitted info | No |
| `correct` | User points out an agent mistake | Yes |
| `elaborate` | User adds new requirements | Yes |
| `ragequit` | User abandons the conversation | Yes |

The effective turn count discounts non-penalty acts: `T_eff = max(1, T - T_user_clarify)`. An agent that solves the task in 3 turns — one of which answers a user confirmation — pays efficiency cost for only 2 effective turns. `r_eff = 1 - 2·(T_eff - 1) / (T_max - 1)`.

> **Hallucination penalty** checks whether recommended product IDs were actually retrieved: \(\text{hall\_rate} = |\{p \in L : p \notin \text{Seen}\}| / \max(|L|, 1)\), \(r_{\text{hall}} = -\text{clip}(\text{hall\_rate}, 0, 1)\). The agent cannot invent product IDs.

Reviewer: Perhaps we should explain these conceptually rather than starting with the terms.

Author: feedback taken, changes incorporated

| **Retrieval noise** `ε_rank(d)` | Fraction of search results replaced by distractors | 0% | 12% | 24% |
| **Out-of-stock rate** `p_oos(d)` | Items that become unavailable mid-episode | 0% | 30% | 50% |

The remaining eight axes control output size, turn budget, input noise, context switches, retrieval depth, order-history depth, policy complexity, and tool budget. The full table is in the [companion technical report](https://github.com/owlgebra-ai/EcomRLVE-Gym).

**How difficulty advances.** Following RLVE, each environment maintains an independent sliding window `[l_i, h_i]`. Episodes sample `d ~ Uniform[l_i, h_i]`. After 32 rollouts at the upper bound, if the agent passes ≥ 90% of them, the window advances by one. A `d_delta = 4` cap keeps the window width at most 5 levels, ensuring the agent always trains near its capability frontier.

---

## Deep dive: Cart Building (E_CART)

Cart building is a good showcase because it requires the full search → inspect → act loop, has a binary ground truth, and introduces a challenge absent from most recommendation benchmarks: **variant selection**.

### The problem

> The generator samples 1–5 target products (scaling with `d`), each potentially requiring a specific variant (USB-C vs Lightning, Matte vs Glossy) and a quantity > 1. The agent must:

Reviewer: ```suggestion
The generator samples 1–5 target products (scaling in difficulty with `d`), each potentially requiring a specific variant (USB-C vs Lightning, Matte vs Glossy) and a quantity > 1. The agent must:
```

We hypothesise — consistent with RLVE's findings — that C8 agents outperform single-environment specialists, even on the specialist's own task.

---

## Early results

We trained Qwen 3 1.7B with DAPO on C1 (product discovery) for 300 steps as an initial viability study.

| | Config |
|---|--------|
| **Base model** | Qwen 3 1.7B |
| **Algorithm** | DAPO (G = 4 rollouts/prompt) |
| **LR** | 1e-5 |
| **Catalog** | 2M products, FAISS index with `thenlper/gte-small` (384-dim) |
| **User sim** | Qwen3.5 9.7B |

![accuracy_10_levels_dots_each_reach (1)](https://cdn-uploads.huggingface.co/production/uploads/6893dd21467f7d2f5f358a95/sGyMSKDOJ4tqiRSgV7AOR.png)

We saw progressive growth in difficulty reached, confirming that adaptive scheduling produces a steady learning signal rather than the saturation (static-low) or starvation (static-high) patterns predicted by the RLVE paper.

---

## Try it yourself

Run an episode directly in your browser — pick an environment, set the difficulty, and watch the agent work:

<script
	type="module"
	src="https://gradio.s3-us-west-2.amazonaws.com/4.36.1/gradio.js"
></script>

> <gradio-app theme_mode="dark" space="owlgebra-ai/EcomRLVE-Gym"></gradio-app>

Reviewer: I don't know how to operate the demo lol. I clicked "Reset Episode" (thought from the text that was what I had to do), then nothing happened for >100s.

Perhaps a screenshot could be useful as well.

Author: have added few pointers as description. Would add detailed How-to to the HF space itself, since it will have active development
